File: frontend/frontend/src/types/chart.ts
export interface Location {
  latitude: number;
  longitude: number;
  altitude: number;
}

export interface PlanetaryPosition {
  longitude: number;
  latitude?: number;
  distance?: number;
  speed?: number;
}

export interface HouseData {
  cusps: number[];
  ascendant: number;
  midheaven: number;
  armc: number;
  vertex: number;
}

export interface Aspect {
  planet1: string;
  planet2: string;
  aspect: string;
  angle: number;
  orb: number;
  is_major: boolean;
  is_applying: boolean;
}

export interface NakshatraData {
  number: number;
  name: string;
  lord: string;
  pada: number;
  degrees_traversed: number;
  total_degrees: number;
}

export interface BirthChart {
  date_time: string;
  location: Location;
  ayanamsa: number;
  house_system: string;
  planetary_positions: Record<string, PlanetaryPosition>;
  houses: HouseData;
  aspects: Aspect[];
  nakshatras: Record<string, NakshatraData>;
}

File: frontend/backend/app/core/models/chart.py
from datetime import datetime
from typing import Dict, List, Optional
from decimal import Decimal
from pydantic import BaseModel, field_validator, ConfigDict

class Location(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    latitude: Decimal
    longitude: Decimal
    altitude: Decimal = Decimal('0')
    
    @field_validator('latitude')
    @classmethod
    def validate_latitude(cls, v):
        if v < -90 or v > 90:
            raise ValueError('Latitude must be between -90 and 90 degrees')
        return v
    
    @field_validator('longitude')
    @classmethod
    def validate_longitude(cls, v):
        if v < -180 or v > 180:
            raise ValueError('Longitude must be between -180 and 180 degrees')
        return v

class PlanetaryPosition(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    longitude: Decimal
    latitude: Optional[Decimal]
    distance: Optional[Decimal]
    speed: Optional[Decimal]

class HouseData(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    cusps: List[Decimal]
    ascendant: Decimal
    midheaven: Decimal
    armc: Decimal
    vertex: Decimal

class Aspect(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    planet1: str
    planet2: str
    aspect: str
    angle: Decimal
    orb: Decimal
    is_major: bool
    is_applying: bool

class NakshatraData(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    number: int
    name: str
    lord: str
    pada: int
    degrees_traversed: Decimal
    total_degrees: Decimal

class BirthChart(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    
    date_time: datetime
    location: Location
    ayanamsa: int = 1
    house_system: str = 'P'
    planetary_positions: Dict[str, PlanetaryPosition]
    houses: HouseData
    aspects: List[Aspect]
    nakshatras: Dict[str, NakshatraData]
